### PR TITLE
add thread safety protection around setenv given apparent glibc behavior

### PR DESCRIPTION
**Pull request by gabemontero**
@openshift/sig-developer-experience fyi / ptal

resulting from recent revelation from @stevekuznetsov on setenv/glibc lack of thread safety

### CODE REVIEW

File: pkg/cmd/util/env.go [hunk 1]
```diff
@@ -31,3 +34,21 @@ func GetEnv(key string) (string, bool) {
 	}
 	return val, true
 }
+
+// ThreadSafeSetEnv provides locking around the call to os.Setenv
+// do an internet search like "golang setenv thread safe" and
```
Comment (stevekuznetsov) [inline pkg/cmd/util/env.go]: It's technically not just Go, just that environment isn't thread safe period.

**Comment by stevekuznetsov** [comment]
Can we actually make this safe at all?

https://rachelbythebay.com/w/2017/01/30/env/

File: tools/genman/gen_man.go [hunk 2]
```diff
@@ -13,6 +13,8 @@ import (
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
 	"k8s.io/kubernetes/cmd/genutils"
+
+	"github.com/openshift/origin/pkg/cmd/util"
```
Comment (adambkaplan) [inline tools/genman/gen_man.go]: nit - group this with the other openshift imports

**Comment by stevekuznetsov** [comment]
No need to churn on this

File: pkg/cmd/util/env.go (hunk 1 above)
Comment (gabemontero) [reply pkg/cmd/util/env.go]: @stevekuznetsov - based on the POSIX and ISO references I found at https://stackoverflow.com/questions/30476600/how-can-getenv-be-thread-safe and that getenv need not be thread safe, that is why I went with only protecting setenv

**Comment by gabemontero** [comment]
bump @bparees @stevekuznetsov with the ci/prow jobs recovered, this is passing all tests.

reaction to my response on not locking getenv?  shall we reapply the lgtm?

**Comment by bparees** [comment]
> reaction to my response on not locking getenv? shall we reapply the lgtm?

my only question on that would be, can getenv segfault *because* a setenv is happening elsewhere?

that is, getenv itself may be threadsafe, with respect to other getenv calls(because it's read only, presumably), but is it threadsafe when used in conjunction w/ setenv or might it catch a corrupt value that's being modified by setenv?

**Comment by gabemontero** [comment]
On Fri, Aug 17, 2018 at 10:03 AM, Ben Parees <notifications@github.com>
wrote:

> reaction to my response on not locking getenv? shall we reapply the lgtm?
>
> my only question on that would be, can getenv segfault *because* a setenv
> is happening elsewhere?
>
> that is, getenv itself may be threadsafe, with respect to other getenv
> calls(because it's read only, presumably), but is it threadsafe when used
> in conjunction w/ setenv or might it catch a corrupt value that's being
> modified by setenv?
>
I had interpreted the posts I referenced that it would be threadsafe, but
data consistency was not guaranteed.  But second opinions on the
interpretations of those various references certainly could be seen a
prudent.

**Comment by bparees** [comment]
you've read up on it more than i have, if you're convinced i'm fine with it.  it's not hard to go back and add it if we decide it's broken later...  technically this is all broken today.

/lgtm

**Comment by gabemontero** [comment]
the end to end test failure was at:

```
# verify remote images can be pulled directly from the local registry
os::log::info "Docker pullthrough"
os::cmd::expect_success "oc import-image --confirm --from=mysql:latest mysql:pullthrough"
os::cmd::expect_success "docker pull ${DOCKER_REGISTRY}/cache/mysql:pullthrough"
```

the docker pull after the import-image did not work.  The stdout/stderr:

```
Standard output from the command:
Trying to pull repository 172.30.1.1:5000/cache/mysql ... 
Pulling repository 172.30.1.1:5000/cache/mysql

Standard error from the command:
Error: image cache/mysql:pullthrough not found
```

@dmage @legionus @bparees - I've seen some flakes opened for this in the past, but they are closed.  Open a new flake?

**Comment by gabemontero** [comment]
@bparees / @dmage / @legionus -  this time the `os::cmd::try_until_success 'oc get imagestreamtags mysql:pullthrough' $((2*TIME_MIN))` after the `os::cmd::expect_success "oc import-image --confirm --from=mysql:latest mysql:pullthrough"` failed.

**Comment by gabemontero** [comment]
disregard my last comment @bparees @dmage @legionus  ... checked the same failed run

File: pkg/cmd/util/env.go [hunk 3]
```diff
@@ -31,3 +34,21 @@ func GetEnv(key string) (string, bool) {
...
+// read about the glibc stuff
+func ThreadSafeSetEnv(key, value string) error {
+	lock.Lock()
+	defer lock.Unlock()
+	return os.Setenv(key, value)
```
Comment (dmage) [inline pkg/cmd/util/env.go]: /hold

Why do need this? We use Go.

`os.Setenv` is just an alias for `syscall.Setenv`. `syscall.Setenv` already has a mutex: https://golang.org/src/syscall/env_unix.go?s=1927:1963#L83

Comment (gabemontero) [reply pkg/cmd/util/env.go]: Ha .... son of a gun 

Good catch @dmage ... thanks

@stevekuznetsov @bparees fyi ^^^

big false alarm

**Pull request closed without merging**